File: backend/internal/views/leaderboard.go
package views

import (
	"time"

	"github.com/mongj/aoc-leaderboard/internal/models"
)

type MemberView struct {
	LocalScore        int                              `json:"local_score"`
	Stars             int                              `json:"stars"`
	GlobalScore       int                              `json:"global_score"`
	LastStarTimestamp int64                            `json:"last_star_ts"`
	ID                int                              `json:"id"`
	Name              *string                          `json:"name,omitempty"`
	SocialsLink       *string                          `json:"socials_link,omitempty"`
	CompletionDays    map[string]*models.CompletionDay `json:"completion_day_level"`
}

type LeaderboardViews struct {
	Event       string       `json:"event"`
	Members     []MemberView `json:"members"`
	LastUpdated time.Time    `json:"last_updated"`
}

type LeaderboardListViews struct {
	Id         int    `json:"id"`
	InviteCode string `json:"invite_code"`
	Size       int    `json:"size"`
}

File: frontend/src/types/leaderboard.ts
import { Nullable } from './utility';

export interface CompletionLevel {
  star_index: number;
  get_star_ts: number;
}

export type DayCompletion = {
  [part in DayPart]: CompletionLevel;
};

export type DayPart = '1' | '2';

export interface CompletionDayLevel {
  [day: string]: DayCompletion;
}

export interface Member {
  local_score: number;
  global_score: number;
  last_star_ts: number;
  name: Nullable<string>;
  socials_link?: string;
  completion_day_level: CompletionDayLevel;
  id: number;
  stars: number;
}

export interface LeaderboardData {
  members: Member[];
  event: string;
  last_updated: string;
}
export interface LeaderboardListView {
  id: number;
  invite_code: string;
  size: number;
}
